### PR TITLE
Removed Id Usages from Lineage Classes

### PR DESCRIPTION
**Pull request by knazir**
Phasing deprecated `Id` class usages out of lineage-related classes.

### CODE REVIEW

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/LineageWriterDataFabricFacade.java [hunk 1]
```diff
@@ -135,7 +136,8 @@ public StreamConsumer createStreamConsumer(Id.Stream streamName, ConsumerConfig
     datasetCache.addExtraTransactionAware(consumer);
 
     if (programContext.getRun() != null) {
-      lineageWriter.addAccess(programContext.getRun(), streamName, AccessType.READ, programContext.getComponentId());
+      lineageWriter.addAccess(programContext.getRun().toEntityId(), streamName.toEntityId(), AccessType.READ,
+                              (NamespacedEntityId) programContext.getComponentId().toEntityId());
```
Comment (bdmogal) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/LineageWriterDataFabricFacade.java]: what happens when programContext.getComponentId() is null?

File: cdap-app-fabric/src/test/java/co/cask/cdap/metadata/LineageAdminTest.java [hunk 2]
```diff
@@ -518,9 +518,10 @@ public void testScanRange() {
     Assert.assertEquals(101, scanRange.getEnd());
   }
 
-  private void addRuns(Store store, Id.Run... runs) {
-    for (Id.Run run : runs) {
-      store.setStart(run.getProgram(), run.getId(), RunIds.getTime(RunIds.fromString(run.getId()), TimeUnit.SECONDS));
+  private void addRuns(Store store, ProgramRunId... runs) {
+    for (ProgramRunId run : runs) {
+      store.setStart(run.getParent().toId(), run.getEntityName(), RunIds.getTime(RunIds.fromString(run.getEntityName()),
+                                                                           TimeUnit.SECONDS));
```
Comment (bdmogal) [inline cdap-app-fabric/src/test/java/co/cask/cdap/metadata/LineageAdminTest.java]: indentation

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/lineage/LineageDataset.java [hunk 3]
```diff
@@ -157,9 +162,9 @@ public void addAccess(Id.Run run, Id.Stream stream,
           LOG.trace("Got row key = {}", Bytes.toString(row.getRow()));
         }
         RowKey rowKey = parseRow(row);
-        if (run.getId().equals(rowKey.getRunId().getId())) {
-          recordBuilder.add(rowKey.getProgram().toEntityId());
-          recordBuilder.add((NamespacedEntityId) rowKey.getData().toEntityId());
+        if (run.getEntityName().equals(rowKey.getRunId().getId())) {
```
Comment (bdmogal) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/lineage/LineageDataset.java:165]: shouldn't rowKey.getRunId return `ProgramRunId`?

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/lineage/LineageDataset.java [hunk 4]
```diff
@@ -386,79 +391,79 @@ private RowKey parseRow(Row row) {
...
+        return new ProgramId(splitter.getString(), splitter.getString(),
                                ProgramType.valueOfCategoryName(splitter.getString()),
                                splitter.getString());
 
       default: throw new IllegalStateException("Invalid row with marker " +  marker);
     }
   }
 
-  private Id.NamespacedId toComponent(MDSKey.Splitter splitter, Id.Program program) {
+  private NamespacedEntityId toComponent(MDSKey.Splitter splitter, ProgramId program) {
```
Comment (bdmogal) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/lineage/LineageDataset.java:460]: not your change, but can you mark this method as `@Nullable`?

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/writer/LineageWriterDatasetFramework.java [hunk 5]
```diff
@@ -191,8 +192,10 @@ private void doWriteLineage(Id.DatasetInstance datasetInstanceId, AccessType acc
     Id.Run programRunId = programContext.getRun();
     if (programRunId != null) {
       Id.NamespacedId componentId = programContext.getComponentId();
```
Comment (bdmogal) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/writer/LineageWriterDatasetFramework.java]: should `ProgramContext.getComponentId()` be changed to return a `NamespacedEntityId` in this PR itself? That way, you wouldn't need the following null check

File: cdap-proto/src/main/java/co/cask/cdap/proto/id/ProgramRunId.java [hunk 6]
```diff
@@ -56,7 +56,7 @@ public String getProgram() {
 
   @Override
   public String getEntityName() {
-    return getProgram();
+    return run;
```
Comment (bdmogal) [inline cdap-proto/src/main/java/co/cask/cdap/proto/id/ProgramRunId.java:59]: :-)

Comment (knazir) [reply cdap-proto/src/main/java/co/cask/cdap/proto/id/ProgramRunId.java:59]: @riwazp7 🎱

**Comment by bdmogal** [comment]
@knazir mostly looks good, some minor comments. However, let's ensure that upgrade does not break after this change.

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/lineage/LineageDataset.java (hunk 3 above)
Comment (knazir) [reply cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/lineage/LineageDataset.java:165]: @bdmogal it currently returns a `RunId` from `org.apache.twill.api.RunId` should it be changed to return `ProgramRunId` ?

Comment (bdmogal) [reply cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/lineage/LineageDataset.java:165]: no let's keep it as-is right now.

**Comment by knazir** [comment]
Guild is breen: http://builds.cask.co/browse/CDAP-DUT4803-2 going to try upgrade tests before merging.

File: cdap-app-fabric/src/main/java/co/cask/cdap/app/stream/DefaultStreamWriter.java [hunk 7]
```diff
@@ -202,6 +202,6 @@ private void registerStream(Id.Stream stream) {
     }
 
     // Lineage writer handles duplicate accesses internally
-    lineageWriter.addAccess(run, stream, AccessType.WRITE);
+    lineageWriter.addAccess(run.toEntityId(), stream.toEntityId(), AccessType.WRITE);
```
Comment (anwar6953) [inline cdap-app-fabric/src/main/java/co/cask/cdap/app/stream/DefaultStreamWriter.java]: q

File: cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/meta/RemoteLineageWriterHandler.java [hunk 8]
```diff
@@ -19,7 +19,10 @@
 import co.cask.cdap.common.internal.remote.MethodArgument;
 import co.cask.cdap.data2.metadata.lineage.AccessType;
```
Comment (anwar6953) [inline cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/meta/RemoteLineageWriterHandler.java]: q

**Comment by knazir** [comment]
Upgrade tests passed, going to squash and merge.

**Pull request merged**